docs/rfcs/2025-06-20-repartition.md

```diff
@@ -0,0 +1,154 @@
+- Lock the table key
+- For each repartition subprocedure
+  - Stop compaction and snapshot
+  - Forbid new ingestion requests, update metadata, allow ingestion requests.
+  - Update region rule to regions
+  - Pick one region to calculate new manifest for all regions in this repartition group
+  - Let that region to apply new manifest to each region via `RegionEdit`
+  - If failed after some retries, revert this manifest change to other succeeded regions and mark this failure.
+  - If all succeeded, acknowledge those staged version changes and make them visible.
+  - Return result
+- Collect results from subprocedure.
+  - For those who failed, we need to restart those regions to force reconstruct their status from manifests
+  - For those who succeeded, collect and merge their rule diff
+- Unlock the table key
+- Report the result to user.
+- <in background> Reload cache
+- <in background> Maybe trigger a special compaction
+
+In addition of sequential step, rollback is also an important part of this procedure. There are three steps can be rolled back when unrecoverable failure occurs.
+
+If the metadata update is not committed, we can overwrite the metadata to previous version. This step is scoped in the "no ingestion" period, so no new data will be ingested and the status of both datanode and metasrv will be consistent.
+
+If the `RegionEdit` to other regions is not acknowledged, or partial acknowledged, we can directly overwrite the manifest on object storage from the central region (who computes the new manifest), and force region server to reload corresponding region to load its state from object storage to recover.
+
+If the staged version changes are not acknowledged, we can re-compute manifest based on old rule for staged data, and apply them directly like above. This is like another smaller repartition for those staged data.
+
+## Region rule validation and diff calculation
+
+In the current codebase, the rule checker is not complete. It can't check uniqueness and completeness of the rule. This RFC also propose a new way to validate the rule.
+
+The proposed validation way is based on a check-point system, which first generates a group of check-points from the rule, and then check if all the point is covered and only covered by one rule.
+
+All the partition rule expressionis limited to be the form of `<column> <operator> <value>`, and the operator is limited to be comparison operators. Those expressions are allowed to be nested with `AND` and `OR` operators. Based on this, we can first extract all the unique values on each column, adding and subtracting a little epsilon to cover its left and right boundary.
+
+Since we accept integer, float and string as the value type, compute on them directly is not convenient. So we'll first normalize them to a common type and only need to preserve the relative partial ordering. This also avoids the problem of "what is next/previous value" of string and "what's a good precision" for float.
```

Comment:
One false positive is that we consider there are possible values between two adjacent integers, though the column type is integer.

However, a potential benefit is that the rule remains correct after we modify the column type to double......